src/components/Waveform.tsx

```diff
@@ -2,73 +2,134 @@ import { useStore } from "@/src/types/StoreContext";
 import { INITIAL_PIXELS_PER_SECOND } from "@/src/utils/time";
 import { Box } from "@chakra-ui/react";
 import { observer } from "mobx-react-lite";
-import { useRef, useEffect } from "react";
+import { useRef, useEffect, useState } from "react";
+import * as THREE from "three";
+import waveformShaderSrc from "@/src/patterns/shaders/waveform.frag"
+import vert from "@/src/patterns/shaders/default.vert"
+import { Texture } from "@react-three/postprocessing";
 
 export default observer(function Waveform() {
   const initialized = useRef(false);
-  const wavesurferRef = useRef<WaveSurfer | null>(null);
-  const waveformRef = useRef(null);
+  const canvasRef = useRef<HTMLCanvasElement>(null);
 
   const { timer, uiStore } = useStore();
+  const [ numSamples, setNumSamples] = useState(0);
+  
+  class AudioTextureStore {
+    texture?: THREE.DataTexture;
 
+    setTexture(texture: THREE.DataTexture) {
+      this.texture = texture;
+    }
+  }
+
+  const audioTextureStore = new AudioTextureStore();
+  
   useEffect(() => {
     if (initialized.current) return;
-
     const create = async () => {
-      // Cannot be run on the server, so we need to use dynamic import
-      const WaveSurfer = (await import("wavesurfer.js")).default;
+      const audioContext = new AudioContext();
+      fetch("/cloudkicker-explorebecurious.mp3")
+        .then(response => response.arrayBuffer())
+        .then(arrayBuffer => audioContext.decodeAudioData(arrayBuffer))
+        .then(audioBuffer => {
+          const audioData = audioBuffer.getChannelData(0);
+          const floatArray = Float32Array.from(audioData);
+          // console.log(floatArray.length);
+          // const sampleTex = new THREE.DataTexture(
+          //   floatArray, 
+          //   audioBuffer.length, 
+          //   1, 
+          //   THREE.RedFormat, 
+          //   THREE.FloatType
+          // );
+          // sampleTex.needsUpdate = true;
 
-      // https://wavesurfer-js.org/docs/options.html
-      const options = {
-        container: waveformRef.current!,
-        waveColor: "#eee",
-        progressColor: "#0178FF",
-        cursorColor: "#00000000",
-        barWidth: 2,
-        barRadius: 2,
-        responsive: true,
-        height: 32,
-        normalize: true,
-        partialRender: true,
-        hideScrollbar: true,
-        interact: false,
-        minPxPerSec: INITIAL_PIXELS_PER_SECOND,
-      };
-      wavesurferRef.current = WaveSurfer.create(options);
-      wavesurferRef.current.zoom(INITIAL_PIXELS_PER_SECOND);
+          const sampleCount = audioBuffer.length;
+          const sampleTexWidth = 1024;
+          const sampleTexHeight = Math.floor(sampleCount / (4 * sampleTexWidth));
+          const sampleTex = new THREE.DataTexture(floatArray, sampleTexWidth, sampleTexHeight, THREE.RGBAFormat, THREE.FloatType);
```

Comment:
Unfortunately, this is the only way to pass the required sample data to the shader. Using a plain float array doesn't work for several reasons:

1. Arrays can't be dynamically sized. You can have a `float[16] samples`, but you can't have a `float[] samples`.

Now, you are no doubt thinking, "ah but can't we just preallocate a very large array, large enough for any possible song, and then pass in the number of samples to read from it?" And unfortunately,

2. Arrays can't be very big. `float[[1024]` is about what you're working with. The Cloudkicker song has ~15M samples.

So, data textures are the only real solution. Even those are pushing it, the Cloudkicker track ends up being 1024 x 3900, which is pushing 4k texture size, a common HW limit. Some very modern HW supports 8k+ texture resolution, but, you know, #compatibility.

Reply:
> Now, you are no doubt thinking, "ah but can't we just preallocate a very large array, large enough for any possible song, and then pass in the number of samples to read from it?"

I was thinking that, get out of my head!